core/src/main/java/org/elasticsearch/search/SearchService.java

```diff
@@ -147,6 +148,8 @@
 
     private final ConcurrentMapLong<SearchContext> activeContexts = ConcurrentCollections.newConcurrentMapLongWithAggressiveConcurrency();
 
+    private final CopyOnWriteArrayList<Long> releasedContexts = new CopyOnWriteArrayList<>();
```

Comment:
I have try to find a way to mark contexts for the reason if there are coming requests for contexts we have already released.